File: ml-controller/services/market_regime_state.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Protocol


MARKET_REGIME_STATE_KEY = "market_regime_state"
LEGACY_REGIME_KEY = "ml:regime"
LEGACY_REGIME_META_KEY = "ml:regime:meta"


class RegimeKVReader(Protocol):
    def get(self, key: str) -> str | None: ...
    def get_json(self, key: str, default: Any = None) -> Any: ...


def normalize_regime_label(raw: Any) -> tuple[str, str] | None:
    text = str(raw or "").strip().lower()
    if not text:
        return None
    if text.startswith("bull"):
        return "bull_market", "bull"
    if text.startswith("bear"):
        return "bear_market", "bear"
    if text.startswith("volatile"):
        return "volatile", "volatile"
    if text.startswith("sideway"):
        return "sideways", "sideways"
    return None


def _as_float_map(raw: Any) -> dict[str, float]:
    if not isinstance(raw, dict):
        return {}
    out: dict[str, float] = {}
    for key, value in raw.items():
        try:
            parsed = float(value)
        except (TypeError, ValueError):
            continue
        if parsed >= 0:
            out[str(key)] = parsed
    return out


def _contract_from_state(state: dict[str, Any], *, source: str) -> dict[str, Any] | None:
    normalized = normalize_regime_label(state.get("label") or state.get("regime") or state.get("family"))
    if not normalized:
        return None
    label, family = normalized
    regime_surface = _as_float_map(
        state.get("regime_surface")
        or state.get("regime_probabilities")
        or state.get("probabilities")
        or {}
    )
    if not regime_surface:
        regime_surface = {label: 1.0}
    return {
        "schema_version": "market-regime-state-v1",
        "label": label,
        "raw_label": state.get("raw_label") or label,
        "alpha_regime": family,
        "family": family,
        "source": source,
        "run_date": state.get("run_date"),
        "computed_at": state.get("computed_at") or state.get("pushed_at") or datetime.now(timezone.utc).isoformat(),
        "regime_surface": regime_surface,
        "regime_index": state.get("regime_index"),
        "hmm_state": state.get("hmm_state"),
        "label_zh": state.get("label_zh"),
        "consensus_threshold": state.get("consensus_threshold"),
        "weight_multipliers": state.get("weight_multipliers") if isinstance(state.get("weight_multipliers"), dict) else {},
        "regime_evidence": state.get("regime_evidence") if isinstance(state.get("regime_evidence"), dict) else {},
        "transition_guard": state.get("transition_guard") if isinstance(state.get("transition_guard"), dict) else {},
        "monitors": state.get("monitors") if isinstance(state.get("monitors"), dict) else {},
        "missing": False,
    }


def resolve_market_regime_contract(kv: RegimeKVReader) -> dict[str, Any]:
    current = kv.get_json(MARKET_REGIME_STATE_KEY, default={}) or {}
    if isinstance(current, dict) and current.get("schema_version") == "market-regime-state-v1":
        contract = _contract_from_state(current, source="market_regime_state")
        if contract:
            return contract

    legacy_meta = kv.get_json(LEGACY_REGIME_META_KEY, default={}) or {}
    if isinstance(legacy_meta, dict):
        contract = _contract_from_state(legacy_meta, source="legacy_meta")
        if contract:
            return contract

    raw_label = kv.get(LEGACY_REGIME_KEY)
    normalized = normalize_regime_label(raw_label)
    if normalized:
        label, family = normalized
        return {
            "schema_version": "market-regime-state-v1",
            "label": label,
            "raw_label": label,
            "alpha_regime": family,
            "family": family,
            "source": "legacy_label",
            "run_date": None,
            "computed_at": datetime.now(timezone.utc).isoformat(),
            "regime_surface": {label: 1.0},
            "regime_index": None,
            "hmm_state": None,
            "label_zh": None,
            "consensus_threshold": None,
            "weight_multipliers": {},
            "regime_evidence": {},
            "transition_guard": {},
            "monitors": {},
            "missing": False,
        }

    return {
        "schema_version": "market-regime-state-v1",
        "label": None,
        "raw_label": None,
        "alpha_regime": "unknown",
        "family": "unknown",
        "source": "missing",
        "run_date": None,
        "computed_at": None,
        "regime_surface": {},
        "regime_index": None,
        "hmm_state": None,
        "label_zh": None,
        "consensus_threshold": None,
        "weight_multipliers": {},
        "regime_evidence": {},
        "transition_guard": {},
        "monitors": {},
        "missing": True,
    }


def build_market_regime_contract_from_market_env(
    market_env: dict[str, Any] | None,
    *,
    run_date: str | None = None,
) -> dict[str, Any]:
    """Build a dated fallback contract from the already-loaded market_env.

    This is only for historical/direct controller reruns where Worker readiness
    can be bypassed and KV market_regime_state is unavailable. The source stays
    explicit so production telemetry does not mistake it for a normal HMM KV
    regime-compute result.
    """
    if not isinstance(market_env, dict) or not market_env:
        return {
            "schema_version": "market-regime-state-v1",
            "label": None,
            "raw_label": None,
            "alpha_regime": "unknown",
            "family": "unknown",
            "source": "market_env_fallback_missing",
            "run_date": run_date,
            "computed_at": None,
            "regime_surface": {},
            "regime_index": None,
            "hmm_state": None,
            "label_zh": None,
            "consensus_threshold": None,
            "weight_multipliers": {},
            "regime_evidence": {},
            "transition_guard": {},
            "monitors": {},
            "missing": True,
        }

    try:
        from services.market_regime_evidence import build_regime_evidence_pack
    except Exception:
        return {
            "schema_version": "market-regime-state-v1",
            "label": None,
            "raw_label": None,
            "alpha_regime": "unknown",
            "family": "unknown",
            "source": "market_env_fallback_unavailable",
            "run_date": run_date,
            "computed_at": None,
            "regime_surface": {},
            "regime_index": None,
            "hmm_state": None,
            "label_zh": None,
            "consensus_threshold": None,
            "weight_multipliers": {},
            "regime_evidence": {},
            "transition_guard": {},
            "monitors": {},
            "missing": True,
        }

    probe = build_regime_evidence_pack(market_env, raw_label="sideways")
    counts = probe.get("support_counts") if isinstance(probe, dict) else {}
    bearish = int((counts or {}).get("bearish") or 0)
    bullish = int((counts or {}).get("bullish") or 0)
    available = int((counts or {}).get("available") or 0)
    if available <= 0:
        return {
            "schema_version": "market-regime-state-v1",
            "label": None,
            "raw_label": None,
            "alpha_regime": "unknown",
            "family": "unknown",
            "source": "market_env_fallback_no_evidence",
            "run_date": run_date,
            "computed_at": None,
            "regime_surface": {},
            "regime_index": None,
            "hmm_state": None,
            "label_zh": None,
            "consensus_threshold": None,
            "weight_multipliers": {},
            "regime_evidence": probe if isinstance(probe, dict) else {},
            "transition_guard": {},
            "monitors": {},
            "missing": True,
        }

    if bearish >= 3 and bearish >= bullish:
        raw_label = "bear_market"
    elif bullish >= 3 and bullish > bearish:
        raw_label = "bull_market"
    elif bearish >= 2:
        raw_label = "volatile"
    else:
        raw_label = "sideways"

    evidence_pack = build_regime_evidence_pack(market_env, raw_label=raw_label)
    label = str(evidence_pack.get("effective_label") or raw_label)
    normalized = normalize_regime_label(label)
    if not normalized:
        normalized = ("sideways", "sideways")
        label = "sideways"
    normalized_label, family = normalized
    surface = {
        "bull_market": 0.0,
        "bear_market": 0.0,
        "volatile": 0.0,
        "sideways": 0.0,
    }
    surface[normalized_label] = 1.0
    return {
        "schema_version": "market-regime-state-v1",
        "label": normalized_label,
        "raw_label": raw_label,
        "alpha_regime": family,
        "family": family,
        "source": "market_env_fallback",
        "run_date": run_date,
        "computed_at": datetime.now(timezone.utc).isoformat(),
        "regime_surface": surface,
        "regime_index": None,
        "hmm_state": None,
        "label_zh": None,
        "consensus_threshold": None,
        "weight_multipliers": {},
        "regime_evidence": evidence_pack,
        "transition_guard": evidence_pack.get("transition_guard") if isinstance(evidence_pack.get("transition_guard"), dict) else {},
        "monitors": evidence_pack.get("monitors") if isinstance(evidence_pack.get("monitors"), dict) else {},
        "missing": False,
    }

File: worker/src/lib/marketRegimeState.ts
export const MARKET_REGIME_STATE_KEY = 'market_regime_state'
export const LEGACY_REGIME_KEY = 'ml:regime'
export const LEGACY_REGIME_META_KEY = 'ml:regime:meta'

export type MarketRegimeFamily = 'bull' | 'bear' | 'volatile' | 'sideways'
export type MarketRegimeLegacyLabel = 'bull_market' | 'bear_market' | 'volatile' | 'sideways'
export type MarketRegimeStateSource = 'hmm' | 'legacy_meta' | 'legacy_label' | 'fallback'

export interface MarketRegimeState {
  schema_version: 'market-regime-state-v1'
  label: MarketRegimeLegacyLabel
  raw_label: MarketRegimeLegacyLabel
  family: MarketRegimeFamily
  run_date: string | null
  computed_at: string
  source: MarketRegimeStateSource
  regime_index: number
  hmm_state: number
  label_zh: string
  regime_surface: Record<string, number>
  consensus_threshold: number
  weight_multipliers: Record<string, number>
  regime_evidence: Record<string, unknown>
  transition_guard: Record<string, unknown>
  monitors: Record<string, unknown>
  downstream_contract: {
    primary_kv_key: typeof MARKET_REGIME_STATE_KEY
    legacy_mirror_keys: [typeof LEGACY_REGIME_KEY, typeof LEGACY_REGIME_META_KEY]
    read_policy: 'market_regime_state_first_legacy_fallback'
    consumers: string[]
  }
  pushed_at?: string
}

export function normalizeRegimeLabel(raw: unknown): { label: MarketRegimeLegacyLabel; family: MarketRegimeFamily } | null {
  const text = String(raw ?? '').trim().toLowerCase()
  if (!text) return null
  if (text.startsWith('bull')) return { label: 'bull_market', family: 'bull' }
  if (text.startsWith('bear')) return { label: 'bear_market', family: 'bear' }
  if (text.startsWith('volatile')) return { label: 'volatile', family: 'volatile' }
  if (text.startsWith('sideway')) return { label: 'sideways', family: 'sideways' }
  return null
}

function asNumber(value: unknown, fallback: number): number {
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : fallback
}

function normalizeSurface(raw: unknown): Record<string, number> {
  if (!raw || typeof raw !== 'object') return {}
  const out: Record<string, number> = {}
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    const parsed = Number(value)
    if (Number.isFinite(parsed) && parsed >= 0) out[key] = parsed
  }
  return out
}

function normalizeNumberMap(raw: unknown): Record<string, number> {
  if (!raw || typeof raw !== 'object') return {}
  const out: Record<string, number> = {}
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    const parsed = Number(value)
    if (Number.isFinite(parsed)) out[key] = parsed
  }
  return out
}

function normalizeObject(raw: unknown): Record<string, unknown> {
  return raw && typeof raw === 'object' && !Array.isArray(raw)
    ? { ...(raw as Record<string, unknown>) }
    : {}
}

function contract(): MarketRegimeState['downstream_contract'] {
  return {
    primary_kv_key: MARKET_REGIME_STATE_KEY,
    legacy_mirror_keys: [LEGACY_REGIME_KEY, LEGACY_REGIME_META_KEY],
    read_policy: 'market_regime_state_first_legacy_fallback',
    consumers: [
      'ml_payload_builder',
      'recommendation_service',
      'alpha_framework',
      'adaptive_params',
      'sltp_overlay',
      'paper_entry',
      'paper_exit',
      'risk_triggers',
    ],
  }
}

export function buildMarketRegimeState(input: {
  label?: unknown
  runDate?: string | null
  computedAt?: string | null
  params?: Record<string, unknown> | null
  source?: MarketRegimeStateSource
}): MarketRegimeState {
  const params = input.params ?? {}
  const normalized = normalizeRegimeLabel(input.label ?? params.label ?? 'sideways') ?? {
    label: 'sideways' as const,
    family: 'sideways' as const,
  }
  const computedAt = String(input.computedAt || params.computed_at || new Date().toISOString())
  const runDate = input.runDate ?? String(params.run_date ?? '')
  const rawLabel = normalizeRegimeLabel(params.raw_label)?.label ?? normalized.label
  const regimeEvidence = normalizeObject(params.regime_evidence)
  return {
    schema_version: 'market-regime-state-v1',
    label: normalized.label,
    raw_label: rawLabel,
    family: normalized.family,
    run_date: runDate || null,
    computed_at: computedAt,
    source: input.source ?? 'hmm',
    regime_index: asNumber(params.regime_index, normalized.label === 'bull_market' ? 0 : normalized.label === 'volatile' ? 1 : normalized.label === 'bear_market' ? 3 : 2),
    hmm_state: asNumber(params.hmm_state, -1),
    label_zh: String(params.label_zh ?? ''),
    regime_surface: normalizeSurface(params.regime_surface ?? params.regime_probabilities ?? params.probabilities),
    consensus_threshold: asNumber(params.consensus_threshold, 0.60),
    weight_multipliers: normalizeNumberMap(params.weight_multipliers),
    regime_evidence: regimeEvidence,
    transition_guard: normalizeObject(params.transition_guard ?? regimeEvidence.transition_guard),
    monitors: normalizeObject(params.monitors ?? regimeEvidence.monitors),
    downstream_contract: contract(),
  }
}

function parseMarketRegimeState(raw: unknown): MarketRegimeState | null {
  if (!raw || typeof raw !== 'object') return null
  const obj = raw as Record<string, unknown>
  if (obj.schema_version !== 'market-regime-state-v1') return null
  const normalized = normalizeRegimeLabel(obj.label)
  if (!normalized) return null
  return {
    schema_version: 'market-regime-state-v1',
    label: normalized.label,
    raw_label: normalizeRegimeLabel(obj.raw_label)?.label ?? normalized.label,
    family: normalized.family,
    run_date: typeof obj.run_date === 'string' && obj.run_date ? obj.run_date : null,
    computed_at: String(obj.computed_at || new Date(0).toISOString()),
    source: (obj.source as MarketRegimeStateSource) || 'fallback',
    regime_index: asNumber(obj.regime_index, 2),
    hmm_state: asNumber(obj.hmm_state, -1),
    label_zh: String(obj.label_zh ?? ''),
    regime_surface: normalizeSurface(obj.regime_surface),
    consensus_threshold: asNumber(obj.consensus_threshold, 0.60),
    weight_multipliers: normalizeNumberMap(obj.weight_multipliers),
    regime_evidence: normalizeObject(obj.regime_evidence),
    transition_guard: normalizeObject(obj.transition_guard),
    monitors: normalizeObject(obj.monitors),
    downstream_contract: contract(),
    pushed_at: typeof obj.pushed_at === 'string' ? obj.pushed_at : undefined,
  }
}

async function readJson(kv: KVNamespace, key: string): Promise<unknown | null> {
  return await kv.get(key, 'json').catch(() => null)
}

export async function readMarketRegimeState(kv: KVNamespace): Promise<MarketRegimeState | null> {
  const current = parseMarketRegimeState(await readJson(kv, MARKET_REGIME_STATE_KEY))
  if (current) return current

  const meta = await readJson(kv, LEGACY_REGIME_META_KEY)
  if (meta && typeof meta === 'object') {
    const legacy = meta as Record<string, unknown>
    const label = normalizeRegimeLabel(legacy.label)
    if (label) {
      return buildMarketRegimeState({
        label: label.label,
        runDate: typeof legacy.run_date === 'string' ? legacy.run_date : null,
        computedAt: typeof legacy.computed_at === 'string' ? legacy.computed_at : typeof legacy.pushed_at === 'string' ? legacy.pushed_at : null,
        params: legacy,
        source: 'legacy_meta',
      })
    }
  }

  const legacyLabel = await kv.get(LEGACY_REGIME_KEY, 'text').catch(() => null)
  const normalized = normalizeRegimeLabel(legacyLabel)
  if (!normalized) return null
  return buildMarketRegimeState({
    label: normalized.label,
    params: {},
    source: 'legacy_label',
  })
}

export async function readCurrentLegacyRegimeLabel(kv: KVNamespace): Promise<MarketRegimeLegacyLabel | null> {
  return (await readMarketRegimeState(kv))?.label ?? null
}

export async function readCurrentRegimeFamily(kv: KVNamespace): Promise<MarketRegimeFamily | null> {
  return (await readMarketRegimeState(kv))?.family ?? null
}

export async function persistMarketRegimeState(
  kv: KVNamespace,
  state: MarketRegimeState,
  options: { expirationTtl?: number } = {},
): Promise<void> {
  const ttl = options.expirationTtl ?? 2 * 86400
  const pushedAt = new Date().toISOString()
  const payload: MarketRegimeState = { ...state, downstream_contract: contract(), pushed_at: pushedAt }
  await kv.put(MARKET_REGIME_STATE_KEY, JSON.stringify(payload), { expirationTtl: ttl })
  await kv.put(LEGACY_REGIME_KEY, payload.label, { expirationTtl: ttl })
  await kv.put(LEGACY_REGIME_META_KEY, JSON.stringify({
    label: payload.label,
    raw_label: payload.raw_label,
    family: payload.family,
    regime_index: payload.regime_index,
    hmm_state: payload.hmm_state,
    label_zh: payload.label_zh,
    regime_surface: payload.regime_surface,
    consensus_threshold: payload.consensus_threshold,
    weight_multipliers: payload.weight_multipliers,
    regime_evidence: payload.regime_evidence,
    transition_guard: payload.transition_guard,
    monitors: payload.monitors,
    run_date: payload.run_date,
    computed_at: payload.computed_at,
    market_regime_state_key: MARKET_REGIME_STATE_KEY,
    pushed_at: pushedAt,
  }), { expirationTtl: ttl })
}
